File: helper-app/internal/platform/mail.go
package platform

import (
	"fmt"
	"os/exec"
	"runtime"
	"strings"
)

// ComposeMail öppnar OS:s mail-app med ett förifyllt kompositions-fönster
// och en bifogad fil. Returnerar utan att vänta på user-action (mail-app:n
// startar — användaren fyller adressat och skickar).
//
//   - macOS: AppleScript mot Mail.app. Skapar "make new outgoing message"
//     med subject + body + attachment, sätter visible:true så fönstret poppar.
//   - Linux: xdg-email --attach + --subject + --body. Funkar med
//     Thunderbird, Evolution, etc. via xdg-utils.
//   - Windows: rundll32 + mailto. Begränsat: bifogning via mailto stöds
//     inte officiellt; Outlook respekterar det informellt, andra klienter
//     ignorerar bilagan. Best-effort.
func ComposeMail(opts ComposeMailOpts) error {
	if opts.AttachmentPath == "" {
		return fmt.Errorf("attachmentPath required")
	}
	switch runtime.GOOS {
	case "darwin":
		return composeMac(opts)
	case "linux":
		return composeLinux(opts)
	case "windows":
		return composeWindows(opts)
	default:
		return fmt.Errorf("unsupported OS: %s", runtime.GOOS)
	}
}

// ComposeMailOpts — input till ComposeMail.
type ComposeMailOpts struct {
	To             string // valfri, default tom
	Subject        string
	Body           string
	AttachmentPath string // absolut sökväg till bilagan
}

func composeMac(opts ComposeMailOpts) error {
	// AppleScript: kapsla in body som "via shell" så multi-line + svenska
	// tecken funkar. Använder POSIX file för bilagan så Mail.app hanterar
	// path:en korrekt.
	script := fmt.Sprintf(`
tell application "Mail"
  activate
  set newMsg to make new outgoing message with properties {visible:true, subject:%s, content:%s}
  tell newMsg
    %s
    tell content
      make new attachment with properties {file name:(POSIX file %s)} at after the last paragraph
    end tell
  end tell
end tell`,
		applescriptQuote(opts.Subject),
		applescriptQuote(opts.Body),
		macToRecipient(opts.To),
		applescriptQuote(opts.AttachmentPath),
	)
	return exec.Command("osascript", "-e", script).Start()
}

func macToRecipient(to string) string {
	if to == "" {
		return ""
	}
	return "make new to recipient at end of to recipients with properties {address:" + applescriptQuote(to) + "}"
}

// applescriptQuote — kapsla en sträng som AppleScript-literal. Escapar
// dubbel-citationstecken och backslash så strängen är säker.
func applescriptQuote(s string) string {
	r := strings.NewReplacer(`\`, `\\`, `"`, `\"`)
	return `"` + r.Replace(s) + `"`
}

func composeLinux(opts ComposeMailOpts) error {
	args := []string{
		"--attach", opts.AttachmentPath,
		"--subject", opts.Subject,
		"--body", opts.Body,
	}
	if opts.To != "" {
		args = append(args, opts.To)
	}
	return exec.Command("xdg-email", args...).Start()
}

func composeWindows(opts ComposeMailOpts) error {
	// Försök 1: Outlook COM via PowerShell — bifoga path:en korrekt.
	// Faller tillbaka på mailto om Outlook inte är installerat.
	ps := fmt.Sprintf(`
$ol = New-Object -ComObject Outlook.Application
$mail = $ol.CreateItem(0)
$mail.To = '%s'
$mail.Subject = '%s'
$mail.Body = '%s'
$mail.Attachments.Add('%s') | Out-Null
$mail.Display()`,
		escapePs(opts.To), escapePs(opts.Subject), escapePs(opts.Body), escapePs(opts.AttachmentPath))
	cmd := exec.Command("powershell", "-NoProfile", "-Command", ps)
	if err := cmd.Start(); err == nil {
		return nil
	}
	// Fallback: mailto (utan bilaga — Windows stöder inte standard-attach)
	url := fmt.Sprintf("mailto:%s?subject=%s&body=%s",
		opts.To, escapeURL(opts.Subject), escapeURL(opts.Body))
	return exec.Command("rundll32", "url.dll,FileProtocolHandler", url).Start()
}

func escapePs(s string) string {
	return strings.ReplaceAll(s, "'", "''")
}

func escapeURL(s string) string {
	return strings.ReplaceAll(strings.ReplaceAll(s, " ", "%20"), "\n", "%0A")
}

File: helper-app/src/platform/mail.ts
/**
 * `composeMail` — öppna OS:ets mail-app med ett förifyllt kompositions-
 * fönster + bifogad fil. Returnerar utan att vänta på user-action.
 * (Port av Go:s platform.ComposeMail.)
 *
 *   - macOS: AppleScript mot Mail.app (make new outgoing message, visible).
 *   - Linux: `xdg-email --attach --subject --body` (Thunderbird/Evolution …).
 *   - Windows: Outlook COM via PowerShell, fallback `mailto:` (utan bilaga).
 *
 * Kommando-byggarna (`mailCommand`, `windowsFallbackCommand` + script-/
 * arg-/url-byggarna) är rena → testbara utan spawn (SOLID).
 */

import type { Command } from "./command.ts";
import { currentPlatform, type Platform } from "./runtime.ts";
import { spawnDetached } from "./spawn.ts";

export interface ComposeMailOpts {
  /** Valfri, default tom. */
  to?: string;
  subject: string;
  body: string;
  /** Absolut sökväg till bilagan. */
  attachmentPath: string;
}

export async function composeMail(opts: ComposeMailOpts): Promise<void> {
  if (opts.attachmentPath === "") {
    throw new Error("attachmentPath required");
  }
  const platform = currentPlatform();
  if (platform === "windows") {
    await composeWindows(opts);
    return;
  }
  const { cmd, args } = mailCommand(platform, opts);
  await spawnDetached(cmd, args).started;
}

/** Primärt mail-kommando per plattform (ren). Kastar för okänt OS. */
export function mailCommand(platform: Platform, opts: ComposeMailOpts): Command {
  switch (platform) {
    case "darwin":
      return { cmd: "osascript", args: ["-e", macScript(opts)] };
    case "linux":
      return { cmd: "xdg-email", args: linuxArgs(opts) };
    case "windows":
      return { cmd: "powershell", args: ["-NoProfile", "-Command", windowsScript(opts)] };
    default:
      throw new Error(`unsupported OS: ${platform}`);
  }
}

/** Windows-fallback: mailto (utan bilaga — stöds inte av standard-attach). */
export function windowsFallbackCommand(opts: ComposeMailOpts): Command {
  const url = `mailto:${opts.to ?? ""}?subject=${escapeUrl(opts.subject)}&body=${escapeUrl(opts.body)}`;
  return { cmd: "rundll32", args: ["url.dll,FileProtocolHandler", url] };
}

async function composeWindows(opts: ComposeMailOpts): Promise<void> {
  const primary = mailCommand("windows", opts);
  try {
    await spawnDetached(primary.cmd, primary.args).started;
  } catch {
    const fb = windowsFallbackCommand(opts);
    await spawnDetached(fb.cmd, fb.args).started;
  }
}

export function macScript(opts: ComposeMailOpts): string {
  return `
tell application "Mail"
  activate
  set newMsg to make new outgoing message with properties {visible:true, subject:${applescriptQuote(opts.subject)}, content:${applescriptQuote(opts.body)}}
  tell newMsg
    ${macToRecipient(opts.to ?? "")}
    tell content
      make new attachment with properties {file name:(POSIX file ${applescriptQuote(opts.attachmentPath)})} at after the last paragraph
    end tell
  end tell
end tell`;
}

export function macToRecipient(to: string): string {
  if (to === "") return "";
  return `make new to recipient at end of to recipients with properties {address:${applescriptQuote(to)}}`;
}

/** Kapsla en sträng som AppleScript-literal (escapar `\` och `"`). */
export function applescriptQuote(s: string): string {
  return `"${s.replace(/\\/g, "\\\\").replace(/"/g, '\\"')}"`;
}

export function linuxArgs(opts: ComposeMailOpts): string[] {
  const args = ["--attach", opts.attachmentPath, "--subject", opts.subject, "--body", opts.body];
  if (opts.to !== undefined && opts.to !== "") args.push(opts.to);
  return args;
}

export function windowsScript(opts: ComposeMailOpts): string {
  return [
    "$ol = New-Object -ComObject Outlook.Application",
    "$mail = $ol.CreateItem(0)",
    `$mail.To = '${escapePs(opts.to ?? "")}'`,
    `$mail.Subject = '${escapePs(opts.subject)}'`,
    `$mail.Body = '${escapePs(opts.body)}'`,
    `$mail.Attachments.Add('${escapePs(opts.attachmentPath)}') | Out-Null`,
    "$mail.Display()",
  ].join("\n");
}

/** PowerShell single-quote-escape (`'` → `''`). */
export function escapePs(s: string): string {
  return s.replace(/'/g, "''");
}

/** Minimal mailto-escape (mellanslag + radbryt). */
export function escapeUrl(s: string): string {
  return s.replace(/ /g, "%20").replace(/\n/g, "%0A");
}
